### PR TITLE
Remove STATIC option from add_library and export DLL symbols on MSVC

### LINKED ISSUE #38: Library is forcefully built as static
https://github.com/sammycage/lunasvg/blob/5f4d0388c1259a6acc884cab90fc655ec5fd95a1/CMakeLists.txt#L10

Is there any reason for forcing the library to always be statically linked? Wouldn't it be a better idea to just use [`BUILD_SHARED_LIBS`](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) and allow the developer to choose?

### PR DESCRIPTION
**Pull request by LHLaurini**
closes #38 

Seems to be working. Only thing is I get warning [C4251](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4251?view=msvc-160), which pretty much just means "be careful when mixing DLLs".

### CODE REVIEW

**Comment by sammycage** [comment]
Thank you ... Testing in progress-------- Original Message --------Subject: [sammycage/lunasvg] Remove STATIC option from add_library and export DLL symbols on Windows (#39)From: Luiz Henrique Laurini To: sammycage/lunasvg CC: Subscribed closes #38
Seems to be working. Only thing is I get warning C4251, which pretty much just means "be careful when mixing DLLs".



You can view, comment on, or merge this pull request online at:
  https://github.com/sammycage/lunasvg/pull/39

Commit Summary

  Don't force library to be built statically
  Add some stuff for building as shared on Windows


File Changes

  
    M
    CMakeLists.txt
    (8)
  
  
    M
    include/document.h
    (20)
  


Patch Links:

  https://github.com/sammycage/lunasvg/pull/39.patch
  https://github.com/sammycage/lunasvg/pull/39.diff


—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/sammycage/lunasvg/pull/39",
"url": "https://github.com/sammycage/lunasvg/pull/39",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by sammycage** [comment]
Thank you ... Testing in progress

File: include/document.h [hunk 1]
```diff
@@ -4,9 +4,19 @@
 #include <memory>
 #include <string>
 
+#if defined(_MSC_VER) && defined(LUNASVG_SHARED)
```
Comment (sammycage) [inline include/document.h:7]: What about other compilers?

Comment (LHLaurini) [reply include/document.h:7]: Mingw doesn't seem to need `__declspec(dllexport)`, unless I did something wrong. Let me make sure.

Comment (LHLaurini) [reply include/document.h:7]: Confirmed. Tested with mingw-w64-gcc 10.2.0. Seems like every symbol is exported, just like on Linux.

**Comment by sammycage** [comment]
Since we are building a shared library, we need to export the target and create a proper config package at ${CMAKE_INSTALL_PREFIX}/lib/cmake/lunasvg with install(EXPORT ...) so other projects can easily import the library using find_package().

**Comment by LHLaurini** [comment]
I agree, but ideally, libraries should always do that, no matter whether they're built as shared or static. I'm willing to implement that too, but probably won't be able to do so this week.

Should I do that in this pull request, or should I open another (since it's technically unrelated)?

**Comment by sammycage** [comment]
> Should I do that in this pull request, or should I open another (since it's technically unrelated)?

You can open another Pull request.

> I agree, but ideally, libraries should always do that, no matter whether they're built as shared or static. I'm willing to implement that too.

If there is anything else you want to add or remove, Please do so.

Thank you for your help <3 <3

File: include/document.h [hunk 2]
```diff
@@ -4,9 +4,19 @@
 #include <memory>
 #include <string>
 
+#if defined(_MSC_VER) && defined(LUNASVG_SHARED)
+# ifdef LUNASVG_EXPORT
+#  define DLLEXPORT __declspec(dllexport)
+# else
+#  define DLLEXPORT __declspec(dllimport)
```
Comment (jardik) [inline include/document.h:11]: I suggest giving these less generic names. Generic names like DLLEXPORT could conflict with other libraries. Something like LUNASVG_DLLEXPORT?

File: include/document.h [hunk 3]
```diff
@@ -199,4 +209,6 @@ class Document
 
 } //namespace lunasvg
 
+#undef LUNASVG_DLLEXPORT
```
Comment (sammycage) [inline include/document.h:212]: You can remove this now

**Pull request merged**